src/bif/map.rs

```diff
@@ -0,0 +1,281 @@
+use crate::atom;
+use crate::immix::Heap;
+use crate::bif::BifResult;
+use crate::exception::{Exception, Reason};
+use crate::process::RcProcess;
+use crate::value::{self, Value};
+use crate::vm;
+use hamt_rs::HamtMap;
+
+pub fn bif_maps_find_2(_vm: &vm::Machine, _process: &RcProcess, args: &[Value]) -> BifResult {
+    unimplemented!();
+}
+
+pub fn bif_maps_get_2(_vm: &vm::Machine, _process: &RcProcess, args: &[Value]) -> BifResult {
+    let map = &args[0];
+    if let Value::Map(m) = map {
+        let hamt_map = &m.0;
+        let target = &args[1];
+        match hamt_map.find(target) {
+            Some(value) => {
+                return Ok((*value).clone());
+            },
+            _ => {
+                let heap = &Heap::new();
+                let tuple = tup2!(heap, Value::Atom(atom::from_str("badkey")), (*target).clone());
+                return Err(Exception::with_value(Reason::EXC_BADARG, tuple));
+            }
+        };
+    }
+    let heap = &Heap::new();
+    let tuple = tup2!(heap, Value::Atom(atom::from_str("badmap")), (*map).clone());
+    Err(Exception::with_value(Reason::EXC_BADARG, tuple))
+}
+
+pub fn bif_maps_from_list_1(_vm: &vm::Machine, _process: &RcProcess, args: &[Value]) -> BifResult {
+    unimplemented!();
+}
+
+pub fn bif_maps_is_key_2(_vm: &vm::Machine, _process: &RcProcess, args: &[Value]) -> BifResult {
+    let map = &args[0];
+    if let Value::Map(m) = map {
+        let hamt_map = &m.0;
+        let target = &args[1];
+        match hamt_map.find(target) {
+            Some(value) => {
+                return Ok(Value::boolean(true));
+            },
+            _ => {
+                return Ok(Value::boolean(false));
+            }
+        };
+
+    }
+    let heap = &Heap::new();
+    let tuple = tup2!(heap, Value::Atom(atom::from_str("badmap")), (*map).clone());
```

Comment:
With those constants, how can I return the argument that make the exception?
According to the documentation of ```maps:is_key/2```, the error ```{badkey, Key}``` and ```{badmap, Map}``` will be returned. 
If I use ```Err(Exception::new(Reason::EXC_BADKEY))`` as a returned value, I don't think we can reproduce the same result as expected.
Or I have some misunderstanding on it?

Reply:
Like this:

```rust
Err(Exception::with_value(tup2!(heap, atom!(BADMAP), map.clone())))
```

Reply:
`atom!(BADMAP)` expands to `Value::Atom(atom::BADMAP)`, where `atom::BADMAP` is a constant number that matches the `badmap` atom index in the `ATOMS` table: https://github.com/archSeer/enigma/blob/master/src/atom.rs#L145
https://github.com/archSeer/enigma/blob/master/src/atom.rs#L188

It's basically an optimization, we give commonly used atoms a well known number, so that we can just use constant numbers and don't need to do a table lookup.

Reply:
I see. Thank you! I understand more!